src/x509/asn1/parse.ts

```diff
@@ -0,0 +1,126 @@
+const RE_TIME_SHORT_YEAR = /^(\d{2})(\d{2})(\d{2})(\d{2})(\d{2})(\d{2})Z$/;
+const RE_TIME_LONG_YEAR = /^(\d{4})(\d{2})(\d{2})(\d{2})(\d{2})(\d{2})Z$/;
+
+// Parse a BigInt from the DER-encoded buffer
+// https://learn.microsoft.com/en-us/windows/win32/seccertenroll/about-integer
+export function parseInteger(buf: Buffer): bigint {
+  let pos = 0;
+  const end = buf.length;
+  let val = buf[pos];
+  const neg = val > 0x7f;
+
+  // Consume any padding bytes
+  const pad = neg ? 0xff : 0x00;
+  while (val == pad && ++pos < end) {
+    val = buf[pos];
+  }
+
+  // Calculate remaining bytes to read
+  const len = end - pos;
+
+  if (len === 0) return BigInt(neg ? -1 : 0);
+
+  // Handle two's complement for negative numbers
+  val = neg ? val - 256 : val;
+
+  // Parse remaining bytes
+  let n = BigInt(val);
+  for (let i = pos + 1; i < end; ++i) {
+    n = n * BigInt(256) + BigInt(buf[i]);
+  }
+
+  return n;
+}
+
+// Parse an ASCII string from the DER-encoded buffer
+// https://learn.microsoft.com/en-us/windows/win32/seccertenroll/about-basic-types#boolean
+export function parseStringASCII(buf: Buffer): string {
+  return buf.toString('ascii');
+}
+
+// Parse a Date from the DER-encoded buffer
+// https://www.rfc-editor.org/rfc/rfc5280#section-4.1.2.5.1
+export function parseTime(buf: Buffer, shortYear: boolean): Date {
+  const timeStr = parseStringASCII(buf);
+
+  // Parse the time string into matches - captured groups start at index 1
+  const m = shortYear
+    ? RE_TIME_SHORT_YEAR.exec(timeStr)
+    : RE_TIME_LONG_YEAR.exec(timeStr);
+
+  if (!m) {
+    throw new Error('invalid time');
+  }
+
+  // Normalize to 4-digit year
+  if (shortYear) {
+    let year = Number(m[1]);
+    year += year >= 50 ? 1900 : 2000;
```

Comment:
Presuming this is for legacy/ancient certificates? Maybe add a comment for future selves?

Reply:
Yeah, this is the scheme described in the spec (https://www.rfc-editor.org/rfc/rfc5280#section-4.1.2.5.1). Will add a comment.